### PR TITLE
[RFC][consensus] refactoring of wait_if_possible to make it clearer

### PR DESCRIPTION
**Pull request by mimoo**
## Motivation

The current code for waiting, when we receive a Proposal that might be in the future, is quite complex.

The current approach attempts to simplify the code.

An open question is: can we use timestamps for everything instead of using both `Instant` and `Duration`? This means that `deadline` would need to be a `Duration` instead of an `Instant`.

## Testing

(This is just an idea, haven't tested it)

### CODE REVIEW

File: consensus/src/util/time_service.rs [hunk 1]
```diff
@@ -148,53 +148,43 @@ pub enum WaitingError {
...
-            let duration_to_min_time = min_duration_since_epoch - current_duration_since_epoch;
-            if duration_to_max_time < duration_to_min_time {
-                return Err(WaitingError::MaxWaitExceeded);
-            }
-        }
+    // are we already passed the deadline?
+    let instant_now = Instant::now();
+    if instant_now >= deadline {
+        return Err(WaitingError::WaitFailed {
+            current_duration_since_epoch: 0,
```
Comment (davidiw) [inline consensus/src/util/time_service.rs]: well somewhat redundant, shouldn't this always be equal to tim_service.get_current_timestamp()?

File: consensus/src/util/time_service.rs [hunk 2]
```diff
@@ -148,53 +148,43 @@ pub enum WaitingError {
...
-                current_duration_since_epoch: waited_duration_since_epoch,
-                wait_duration: sleep_duration,
-            })
-        }
-    } else {
-        Ok(WaitingSuccess::NoWaitRequired {
-            current_duration_since_epoch,
-            early_duration: current_duration_since_epoch - min_duration_since_epoch,
-        })
+    // get current timestamp
```
Comment (davidiw) [inline consensus/src/util/time_service.rs]: not a particularly useful comment.

File: consensus/src/util/time_service.rs [hunk 3]
```diff
@@ -148,53 +148,43 @@ pub enum WaitingError {
...
+    let timestamp_now = time_service.get_current_timestamp();
+    // we don't need to wait
+    if timestamp_now > timestamp_then {
+        return Ok(WaitingSuccess::NoWaitRequired {
+            timestamp_now,
+            early_duration: timestamp_now - timestamp_then,
+        });
+    }
+    // we need to wait, but the deadline will be reached first
```
Comment (davidiw) [inline consensus/src/util/time_service.rs]: maybe add a new line above for readability

File: consensus/src/util/time_service.rs [hunk 4]
```diff
@@ -148,53 +148,43 @@ pub enum WaitingError {
...
-                wait_duration: sleep_duration,
-            })
-        }
-    } else {
-        Ok(WaitingSuccess::NoWaitRequired {
-            current_duration_since_epoch,
-            early_duration: current_duration_since_epoch - min_duration_since_epoch,
-        })
+    // get current timestamp
+    let timestamp_now = time_service.get_current_timestamp();
```
Comment (davidiw) [inline consensus/src/util/time_service.rs]: maybe this should be the first line of the function?

File: consensus/src/util/time_service.rs [hunk 5]
```diff
@@ -148,53 +148,43 @@ pub enum WaitingError {
...
+    // we don't need to wait
```
Comment (davidiw) [inline consensus/src/util/time_service.rs]: make new line between each logical block?

File: consensus/src/util/time_service.rs [hunk 6]
```diff
@@ -148,53 +148,43 @@ pub enum WaitingError {
...
+            early_duration: timestamp_now - timestamp_then,
+        });
+    }
+    // we need to wait, but the deadline will be reached first
+    let time_to_timeout = deadline.duration_since(instant_now);
+    let time_to_wait = timestamp_then - timestamp_now + Duration::from_millis(1);
+    if time_to_timeout < time_to_wait {
+        return Err(WaitingError::MaxWaitExceeded);
     }
+    // we wait
```
Comment (davidiw) [inline consensus/src/util/time_service.rs]: maybe new line above?

**Comment by mimoo** [comment]
There is quite a number of failing tests, so I'm not going to attempt to fix any unless you folks think this is a better approach :D

**Comment by davidiw** [comment]
Please re-base and re-request if still valid.

**Pull request closed without merging**